### PR TITLE
Add writing functionality for dataframes

### LINKED ISSUE #20: Write the DF as rds file
Thanks again for the applied change to read the rds files. 
Is it possible to add a new feature to write the df as an rds file too? In this way, there will be a full bridge between R data and Python.

### PR DESCRIPTION
**Pull request by trossi**
## References to issues or other PRs

Closes #20.

## Describe the proposed changes

This PR will add support for writing pandas dataframes. This turned out to be a quite large change to the current Python-to-R conversion functionality. Overview of changes:
* Functions in `rdata.conversion.to_r` reorganized, a class `ConverterFromPythonToR` added to simplify keeping track of references in RData files.
* Functionality for distinguishing R's NA float value from other NaN values added to `rdata.missing`.
* Added more tests on reading and writing dataframes with various dtypes and a mix of NA and NaN values.
* Added `convert_altrep_to_range()` (`rdata.conversion._conversion`) that enable converting compact intseq to range object (e.g. as a dataframe index).
* Unparsing REF and ALTREP added.

## Additional information

The functions in `rdata.missing` could be useful for users to handle NA values in the desired way, e.g., `pd.arrays.FloatingArray(array)` would set all NaNs (including R's NA value) as "missing", but `pd.arrays.FloatingArray(array, is_na(array))` can be used to set only R's NA values as "missing".

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] The code conforms to the style used in this package (checked with [Ruff](https://docs.astral.sh/ruff/))
- [x] The code is fully documented and typed (type-checked with [Mypy](https://mypy-lang.org/))
- [x] I have added thorough tests for the new/changed functionality

### CODE REVIEW

File: rdata/conversion/_conversion.py [hunk 1]
```diff
@@ -820,6 +872,9 @@ def _convert_next(  # noqa: C901, PLR0912, PLR0915
 
             value = None
 
+        elif obj.info.type == parser.RObjectType.ALTREP:
+            value = convert_altrep_to_range(obj)
```
Comment (vnmabus) [inline rdata/conversion/_conversion.py]: I feel that I am missing something here. Altreps should not be present here when `expand_altrep` is `True` (and that is the default). Also, I am not sure that it is a good idea to convert to different Python objects depending on the internal way some R object is stored (AFAIK, altreps are supposed to be an implementation detail).

File: rdata/conversion/_conversion.py [hunk 2]
```diff
@@ -430,7 +482,7 @@ def dataframe_constructor(
             and isinstance(row_names, np.ma.MaskedArray)
             and row_names.mask[0]
         )
-        else tuple(row_names)
+        else row_names
```
Comment (vnmabus) [inline rdata/conversion/_conversion.py:451]: Why this change?

File: rdata/conversion/to_r.py [hunk 3]
```diff
@@ -47,25 +45,93 @@ def __call__(self, data: Any, *, encoding: Encoding) -> RObject: # noqa: ANN401
     2: 0x20300,
     3: 0x30500,
 })
+R_MINIMUM_VERSION_WITH_ENCODING: Final[int] = 3
+R_MINIMUM_VERSION_WITH_ALTREP: Final[int] = 3
+
+
+def convert_pd_array_to_np_array(
+        pd_array: Any,  # noqa: ANN401
```
Comment (vnmabus) [inline rdata/conversion/to_r.py]: I think all Pandas array classes inherit from `ExtensionArray` nowadays.
```suggestion
    pd_array: pd.api.extensions.ExtensionArray,
```

File: rdata/conversion/to_r.py [hunk 4]
```diff
@@ -47,25 +45,93 @@ def __call__(self, data: Any, *, encoding: Encoding) -> RObject: # noqa: ANN401
...
+    """
+    Convert pandas array object to numpy array.
+
+    Args:
+        pd_array: Pandas array.
+
+    Returns:
+        Numpy array.
+    """
+    if isinstance(pd_array, pd.arrays.StringArray):
```
Comment (vnmabus) [inline rdata/conversion/to_r.py:231]: So, is there some reason not to be able to use masked arrays whenever there are NA values, independently of the array type?

File: rdata/conversion/to_r.py [hunk 5]
```diff
@@ -75,269 +141,453 @@ def build_r_object(
...
-    Returns:
-        R object.
+        format_version: File format version.
+        r_version_serialized: R version written as the creator of the object.
     """
-    r_type = RObjectType.SYM
-    r_value = convert_to_r_object(data.encode(encoding), encoding=encoding)
-    return build_r_object(r_type, value=r_value)
+    def __init__(self, *,
+            encoding: Encoding = "utf-8",
```
Comment (vnmabus) [inline rdata/conversion/to_r.py]: I don't know why, but I noticed you are using 8-spaces indentation in parameter lists sometimes. It should be 4-spaces.

File: rdata/conversion/to_r.py [hunk 6]
```diff
@@ -75,269 +141,453 @@ def build_r_object(
...
+        self.encoding = encoding
+        self.format_version = format_version
+        self.r_version_serialized = r_version_serialized
+        self._references: dict[str | None, tuple[int, RObject | None]] \
+            = {None: (0, None)}
+
+        # In test files the order in which dataframe attributes are written varies.
+        # R can read files with attributes in any order, but this variable
+        # is used in tests to change the attribute order to match with the test file.
+        self.df_attr_order: list[str] | None = None
```
Comment (vnmabus) [inline rdata/conversion/to_r.py]: Is it not just possible to write them always in the same order (the one R uses?).

File: rdata/conversion/to_r.py [hunk 7]
```diff
@@ -75,269 +141,453 @@ def build_r_object(
...
+
+            attributes = {
+                "names": np.array(column_names, dtype=np.dtype("U")),
+                "class": "data.frame",
+                "row.names": row_names,
+            }
+            if self.df_attr_order is not None:
+                attributes = {k: attributes[k] for k in self.df_attr_order}
+
+        else:
```
Comment (vnmabus) [inline rdata/conversion/to_r.py:626]: I would have loved some way to provide a dict of conversions of Python classes to S3/S4 objects (the inverse of what it is done in the conversion from R to Python) instead of hardcoding them. But this is something that can always be improved in another PR.

File: rdata/conversion/to_r.py [hunk 8]
```diff
@@ -75,269 +141,453 @@ def build_r_object(
...
+        if attributes is not None:
+            r_attributes = self.convert_to_r_attributes(attributes)
+        else:
+            r_attributes = None
 
-    return RData(versions, extra, r_object)
+        return build_r_object(r_type, value=r_value,
+                              is_object=is_object,
+                              attributes=r_attributes,
+                              tag=tag, gp=gp)
```
Comment (vnmabus) [inline rdata/conversion/to_r.py]: I have seen this at some points in this PR, but in this case is particularly noticeable. In general, for multiline lines I prefer to have the ending parenthesis (or bracket or brace) in its own line, with the same indentation as the line in which it is opened. Also, it is better for each parameter to be in its own line, for improved readability:
```suggestion
        return build_r_object(
            r_type,
            value=r_value,
            is_object=is_object,
            attributes=r_attributes,
            tag=tag,
            gp=gp,
        )
```

File: rdata/missing.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    """
+    if isinstance(array, np.ndarray):
+        dtype = array.dtype
+        na = get_na_value(dtype)
+        if dtype == np.int32:
+            # Use the native dtype for comparison when possible;
+            # slightly faster than the steps below
+            return array == na  # type: ignore [no-any-return]
+        raw_dtype = f"V{array.dtype.itemsize}"
+        return array.view(raw_dtype) == np.array(na).view(raw_dtype)  # type: ignore [no-any-return]
```
Comment (vnmabus) [inline rdata/missing.py]: I am not sure if I understand what is happening here.

File: rdata/missing.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+    if isinstance(array, int):
+        try:
+            return is_na(np.array(array, dtype=np.int32))
+        except OverflowError:
+            return is_na(np.array(array))
```
Comment (vnmabus) [inline rdata/missing.py]: I also think this deserves some explanation.

File: rdata/parser/_parser.py [hunk 11]
```diff
@@ -941,7 +934,10 @@ def parse_R_object(  # noqa: N802, C901, PLR0912, PLR0915
                     info=altrep_info,
                     state=altrep_state,
                 )
-                attributes = altrep_attr
+                if altrep_attr.info.type != RObjectType.NILVALUE:
```
Comment (vnmabus) [inline rdata/parser/_parser.py]: When does this happen?

File: rdata/tests/test_write.py [hunk 12]
```diff
@@ -126,68 +133,118 @@ def test_convert_to_r(fname: str) -> None:
...
-            )
+            new_r_data = converter.convert_to_r_data(py_data, file_type=file_type)
         except NotImplementedError as e:
             pytest.xfail(str(e))
 
-        assert r_data == new_r_data
         assert str(r_data) == str(new_r_data)
+        assert r_data == new_r_data
+
+        # Check futher that the resulting unparsed data is correct to ensure that
```
Comment (vnmabus) [inline rdata/tests/test_write.py]:
```suggestion
        # Check further that the resulting unparsed data is correct to ensure that
```

**Review by vnmabus — CHANGES_REQUESTED** [review]

File: rdata/conversion/to_r.py (hunk 3 above)
Comment (trossi) [reply rdata/conversion/to_r.py]: Thanks!

File: rdata/conversion/to_r.py (hunk 8 above)
Comment (trossi) [reply rdata/conversion/to_r.py]: I agree. This reads much better.

File: rdata/conversion/_conversion.py (hunk 1 above)
Comment (trossi) [reply rdata/conversion/_conversion.py]: This relates to the support for writing dataframes with range indices like `pd.DataFrame(..., index=range(2, 5))`. 

The reason that this change is in this PR is that it allows testing R-to-Python-to-R roundtrip for such a dataframe (in R `data.frame(..., row.names=2:4)`, this is `test_dataframe_range_rownames.rds`). Here row names is compact_intseq altrep, which would get converted to a regular array by default, which would then fail the roundtrip.

But, other than enabling that testing, this change wouldn't really belong to this PR, so it could be removed if you think that's better.

In general, I feel that it could be nicer if altreps would be expanded in the conversion stage instead of the parsing stage. The reason is that then the RData object would be a one-to-one representation of the file contents, which is not the case for altreps at the moment (unless `expand_altrep=False`). I have interpreted compact_intseq altrep to implement the same idea as Python's range object, so it would be useful if users would have an option to choose whether they want this altrep to be converted to a numpy array or a range object, which seems not possible (or easy?) if altreps are expanded already in parsing. This would be larger change though (beyond this PR). What do you think in general?

File: rdata/conversion/_conversion.py (hunk 2 above)
Comment (trossi) [reply rdata/conversion/_conversion.py:451]: The reason was to keep range object as range object instead of expanding it to a tuple of values (relates to the altrep comment).

File: rdata/conversion/to_r.py (hunk 4 above)
Comment (trossi) [reply rdata/conversion/to_r.py:231]: That could be done. I think it would clarify many things, but it would require changes in parsing side too.

The main issue here could be string arrays, where NA values are `None` at the moment, so should parsing of string arrays return a possibly masked array instead? Also in writing, if user has `strings = ["hello", None]`, now this can be converted to valid "string" array like `array = np.array(strings)` (which is of object dtype, requiring special treatment of object arrays). With masked array user would need to do e.g. `array = np.ma.array(data=["" if s is None else s for s in strings], mask=[s is None for s in strings])`, which might be more cumbersome for users to work with. A benefit would be though that this array is of unicode dtype, which would simplify the numpy array handling.

Do you think it would be useful to start using masked arrays in all these cases?  (I feel it could fit better to a different PR though due to changes in parsing?)

File: rdata/conversion/to_r.py (hunk 5 above)
Comment (trossi) [reply rdata/conversion/to_r.py]: Fixed. My editor seemed to prefer these 8-spaces in these contexts. Could ruff be configured to enforce this spacing?

File: rdata/conversion/to_r.py (hunk 6 above)
Comment (trossi) [reply rdata/conversion/to_r.py]: The order seems to vary depending on *something\**. The default order is the same I get consistently with container image `docker.io/r-base:4.4.0` and the test files added in this PR follow that order. However, the existing test files (e.g., `test_dataframe.rds`) have attributes in different order, so I added this option to make R-to-Python-to-R tests pass, see [here](https://github.com/vnmabus/rdata/blob/957887c60a3f322e523d6ae2ac9916f147f57a38/rdata/tests/test_write.py#L141-L147).

But I agree that this hack is ugly. I could remove this hack and regenerate those old test files so that they have the same order. What do you think?

\* R version is not enough to explain the changing order. The file `test_dataframe.rds` seems to be created with R 4.1.3, but if I create files with `docker.io/r-base:4.1.3`, the order is the same as with `docker.io/r-base:4.4.0`. Just guessing, but maybe these attributes are presented as a dictionary in R too and depending on OS-specific library dependency or something else the dictionary gets dumped to a file in different order.

File: rdata/missing.py (hunk 9 above)
Comment (trossi) [reply rdata/missing.py]: I added a comment to the code (and changed void type to unsigned int). The reason is that as R_FLOAT_NA is a NaN, so it follows NaN-logic: `R_FLOAT_NA == R_FLOAT_NA` is False and `R_FLOAT_NA != R_FLOAT_NA` is True, but also `R_FLOAT_NA == np.nan` is False and `R_FLOAT_NA != np.nan` is True, although R_FLOAT_NA and np.nan are different. So, we need to do byte-by-byte comparison of values to distinguish R_FLOAT_NA from other NaN values.

File: rdata/missing.py (hunk 10 above)
Comment (trossi) [reply rdata/missing.py]: I added a comment to the code. Basically, R seems not to support larger integers than 32-bit, so this attempts to convert Python's int (64 bit or larger) to such (or proceed with larger int and fail later).

File: rdata/parser/_parser.py (hunk 11 above)
Comment (trossi) [reply rdata/parser/_parser.py]: The `altrep_attr` seems to be NIL in all test files, so I changed this to raise NotImplementedError if it's not NIL. The issue here was that with the original line of code, the resulting expanded object got NIL attributes, which seems unnecessary and possibly problematic (as the object might be just a plain integer array). This caused also issues in R-to-Python-to-R tests as these NIL attributes got lost in R-to-Python-to-R roundtrip.

File: rdata/conversion/to_r.py (hunk 7 above)
Comment (trossi) [reply rdata/conversion/to_r.py:626]: Good idea! I restructured these conversion to user-definable constructor functions similarly to R-to-Python conversion. I separated also some other conversion logic to functions.

For `pd.Categorical` this was simple, but for `pd.DataFrame` a bit more complex as it may contain `pd.Categorical` or some other more complex objects. For that reason, the constructor functions take also parameter `convert_to_r_object` that points back to `ConverterFromPythonToR.convert_to_r_object` (which then in turn can keep track of references etc).

Do you have a suggestion for naming convention? Now there is `dataframe_constructor` for both R-to-Python and Python-to-R conversions.

**Review by trossi — COMMENTED** [review]
@vnmabus Thank you for the review. I have pushed the changes and left some comments/questions for discussion.

File: rdata/parser/_parser.py (hunk 11 above)
Comment (trossi) [reply rdata/parser/_parser.py]: This change seems to fail [the readthedocs CI job](https://readthedocs.org/projects/rdata/builds/25811248/). I'll have a look what kind of data it encounters.

Comment (trossi) [reply rdata/parser/_parser.py]: The error comes from [this file](https://zenodo.org/records/7425539/files/core_fungal_pathogens.rds). It contains `wrap_real` altrep with `class` attribute:
```
RObjectType.LIST
  tag:
    RObjectType.REF
      reference: 2
        RObjectType.SYM
          value:
            RObjectType.CHAR(gp=64)
              value:
                b'class'

      value:
        None

  value:
    RObjectType.STR
      value:
        RObjectType.CHAR(gp=64)
          value:
            b'Date'
    RObjectType.NILVALUE
      value:
        None
```
It looks like this is not correctly handled in the existing line of code (`attributes = altrep_attr`) as it doesn't set `info.attributes` or `info.object`, so I suppose here could be some code like this instead:
```python
if altrep_attr.info.type != RObjectType.NILVALUE:
    info.object = ... # add a function to return True if class in attributes 
    info.attributes = True
    attributes_read = True
    attributes = altrep_attr
 ```
 What do you think? (I think that change could be better in a separate PR.)

Comment (trossi) [reply rdata/parser/_parser.py]: I pushed a proposal for a fix in #46.

File: rdata/conversion/_conversion.py (hunk 1 above)
Comment (vnmabus) [reply rdata/conversion/_conversion.py]: I think altreps are intended to be an internal representation detail. That is why I made the default to expand them in the parser: I assumed that, except for very specific situations, users would prefer to ignore their existence.

That said, IMHO the default conversion routines should be able to deal with altreps (in case that `expand_altreps`  is `False`). I still think they should convert it to the same object as if they weren't altreps, for consistency.

For me ranges are a very different beast than NumPy arrays, and returning such different kind of objects for the "same" underlying R object would be confusing. Moreover, there is no equivalent to `range` for compact sequences of float.

If we want to convert an array back to an altrep for space savings we could probably try to detect if the array is a sequence (with `np.diff`, a subtraction, and `np.all` or something like that it would be easy, at least for integer dtype). However that would convert to altrep even arrays that were **not** stored as altreps in the original file, for whatever reason. So, the choice would be between no altreps at all when writing, or all possible ones.

For the above reasons I think it is preferrable **not** to consider altreps in the roundtrip comparison (that is, comparing both files with `expand_altreps=True`. Any thoughts on this?

File: rdata/conversion/to_r.py (hunk 7 above)
Comment (vnmabus) [reply rdata/conversion/to_r.py:626]: I have the feeling that constructors should have a way to receive the particular `Converter` used (also for R to Python constructors). Then they could call that method themselves (for writing), inspect special converter options or whatever they need. However, as they are already users of the current constructor interface, we should only pass that parameter when the constructor accepts it.

I think for now we can pass the `Converter` to the Python to R constructors, as they are new, and if necessary we can see how to pass that also to the R to Python constructors in the future.

File: rdata/conversion/to_r.py (hunk 6 above)
Comment (vnmabus) [reply rdata/conversion/to_r.py]: It would be good to know what affects the order.

In absence of that information I would propose to always write them in the most common order, and try to ignore the order of the attributes when comparing the original and created file. If that is not possible, regenerating the old test files is also doable.

File: rdata/conversion/to_r.py (hunk 5 above)
Comment (vnmabus) [reply rdata/conversion/to_r.py]: Is it currently not enforcing this? From the discussion in https://github.com/astral-sh/ruff/issues/8360 I would have expected it to currently enforce 4 spaces in this case.

File: rdata/conversion/to_r.py (hunk 4 above)
Comment (vnmabus) [reply rdata/conversion/to_r.py:231]: I think we should probably used the [new string dtype](https://numpy.org/devdocs/user/basics.strings.html#variable-width-strings) in the future, when available. This already [supports missing data](https://numpy.org/devdocs/user/basics.strings.html#missing-data-support) using sentinel values.

Comment (trossi) [reply rdata/conversion/to_r.py:231]: Good point, I agree. So let's leave string arrays as they are for now.

How about float arrays with missing values? Would it be useful to convert them to masked arrays similarly to integer arrays? (I think that change could go to another PR though as it changes parsing behavior.)

File: rdata/conversion/to_r.py (hunk 7 above)
Comment (trossi) [reply rdata/conversion/to_r.py:626]: I agree it's nicer to pass the `Converter` object and I changed the constructors to do so. Now this parameter is also optional (not needed for `categorical_constructor`).

File: rdata/conversion/to_r.py (hunk 5 above)
Comment (trossi) [reply rdata/conversion/to_r.py]: You are right. I was using only `ruff check` earlier and that doesn't check formatting, but `ruff format` would fix this spacing (and reformat many other lines throughout the codebase). I changed formatting in some other places too according to `ruff format`.

File: rdata/conversion/to_r.py (hunk 6 above)
Comment (trossi) [reply rdata/conversion/to_r.py]: I tested recreating `test_dataframe_v3.rds` with `Rscript -e 'df = data.frame(class=factor(c("a", "b", "b")), value=c(1L, 2L, 3L)); saveRDS(df, file="test_dataframe_v3.rds")` and I get the same default order in Linux with different R versions as well as in Windows with R 4.1.3. So, it remains unclear how it would be possible to create these files with attributes in any other order.

Ignoring the order in testing would require possibly complex (and unnecessary) logic, so I recreated those test files instead and removed the `df_attr_order` hack.

File: rdata/conversion/_conversion.py (hunk 1 above)
Comment (trossi) [reply rdata/conversion/_conversion.py]: I agree with your points on range and that it would be confusing to always convert intseq altrep to a range object as it doesn't behave similarly to numpy array (in contrast to behavior in R). So, I removed this altrep-to-range conversion from R-to-Python conversion.

In Python-to-R conversion side, I also removed the general conversion of range objects, but changed that logic to apply to pd.RangeIndex only, that is, pd.RangeIndex converts to altrep (if possible; step=1) and pd.Index converts to array. I added a few tests for this case as the roundtrip comparison can't reach this case. Do you think this is ok?

> If we want to convert an array back to an altrep for space savings we could probably try to detect if the array is a sequence

I agree that could be done at some point if needed.

> For the above reasons I think it is preferrable not to consider altreps in the roundtrip comparison (that is, comparing both files with expand_altreps=True. Any thoughts on this?

I agree. The R-Python-R roundtrip tests do file checking only for `expand_altreps=False` (which skips all files with altreps as altreps are then unhandled in R-to-Python conversion).

File: rdata/conversion/_conversion.py (hunk 2 above)
Comment (trossi) [reply rdata/conversion/_conversion.py:451]: I'd keep this change even though I removed the altrep-to-range conversion. The reason is that if a user would like to create a custom `altrep_constructor_dict` that maps compact_intseq to a range object, then that range object wouldn't be expanded here but it would be passed to dataframe as such. What do you think?

File: rdata/conversion/_conversion.py (hunk 1 above)
Comment (vnmabus) [reply rdata/conversion/_conversion.py]: > Do you think this is ok?

In principle it is ok to write altreps when possible by default. I am not sure if we should add an option for the `Converter` not to use altreps, for having compatibility with tools that do not understand them.

File: rdata/conversion/_conversion.py (hunk 2 above)
Comment (vnmabus) [reply rdata/conversion/_conversion.py:451]: I would say that such mapping is an error. However, I do not have anything against this particular change, unless the `tuple` call was necessary for some reason.

File: rdata/conversion/to_r.py (hunk 4 above)
Comment (vnmabus) [reply rdata/conversion/to_r.py:231]: I think that it makes sense to do it for float arrays (in a different PR).

File: rdata/_write.py [hunk 13]
```diff
@@ -40,6 +42,7 @@ def write_rds(
         compression: Compression.
         encoding: Encoding to be used for strings within data.
         format_version: File format version.
+        constructor_dict: Dictionary mapping Python types to R classes.
```
Comment (vnmabus) [inline rdata/_write.py]: That is not really true, right? It maps Python classes to functions to convert them to R classes (which is more powerful, as it can choose a different R class depending on the attributes of the object).

File: rdata/conversion/__init__.py [hunk 14]
```diff
@@ -1,4 +1,5 @@
 """Utilities for converting R objects to Python ones."""
```
Comment (vnmabus) [inline rdata/conversion/__init__.py]: This docstring should change.

File: rdata/conversion/__init__.py [hunk 15]
```diff
@@ -25,7 +26,9 @@
     ts_constructor as ts_constructor,
 )
 from .to_r import (
-    build_r_data as build_r_data,
-    convert_to_r_object as convert_to_r_object,
-    convert_to_r_object_for_rda as convert_to_r_object_for_rda,
+    DEFAULT_CONSTRUCTOR_DICT as DEFAULT_CONSTRUCTOR_DICT,
```
Comment (vnmabus) [inline rdata/conversion/__init__.py]: I do not think this should be exposed here (at least with that name) as it can be easily confused with `DEFAULT_CLASS_MAP`. I think it would be better for the users to import the name directly from `to_r`/`to_python` submodules, and share the same nomenclature.

File: rdata/conversion/__init__.py [hunk 16]
```diff
@@ -25,7 +26,9 @@
...
+    DEFAULT_FORMAT_VERSION as DEFAULT_FORMAT_VERSION,
```
Comment (vnmabus) [inline rdata/conversion/__init__.py]: This name is unclear: default for what? If it is the default version using for writing, maybe it should be only exposed in `to_r` (or maybe in the parser?).

File: rdata/conversion/_conversion.py [hunk 17]
```diff
@@ -394,20 +394,38 @@ def convert_array(
...
         if np.issubdtype(source.dtype, np.integer):
-            return pd.Series(source, dtype=pd.Int32Dtype()).array
-
-        if np.issubdtype(source.dtype, np.bool_):
-            return pd.Series(source, dtype=pd.BooleanDtype()).array
+            dtype = pd.Int32Dtype()
+        elif np.issubdtype(source.dtype, np.floating):
+            # We return the numpy array here, which keeps
+            # R_FLOAT_NA, np.nan, and other NaNs as they were originally in the file.
+            # Users can then decide if they prefer to interpret
```
Comment (vnmabus) [inline rdata/conversion/_conversion.py:406]: How can they decide?

File: rdata/conversion/to_r.py [hunk 18]
```diff
@@ -47,25 +63,226 @@ def __call__(self, data: Any, *, encoding: Encoding) -> RObject: # noqa: ANN401
...
+    """
+    Construct R object components from pandas categorical.
+
+    Args:
+        data: Pandas categorical.
+
+    Returns:
+        Components of the R object.
+    """
+    assert isinstance(data, pd.Categorical)
```
Comment (vnmabus) [inline rdata/conversion/to_r.py]: Are these asserts necessary? Don't we already guarantee that?

File: rdata/conversion/to_r.py [hunk 19]
```diff
@@ -47,25 +63,226 @@ def __call__(self, data: Any, *, encoding: Encoding) -> RObject: # noqa: ANN401
...
+            array = pd_array
+        else:
+            array = convert_pd_array_to_np_array(pd_array)
+        r_series = converter.convert_to_r_object(array)
+        r_value.append(r_series)
+
+    index = data.index
+    if isinstance(index, pd.RangeIndex):
+        assert isinstance(index.start, int)
+        if index.start == 1 and index.stop == data.shape[0] + 1 and index.step == 1:
```
Comment (vnmabus) [inline rdata/conversion/to_r.py:128]: What is happening here?

File: rdata/conversion/to_r.py [hunk 20]
```diff
@@ -24,18 +27,31 @@
...
+            """Convert string to R symbol."""
 
-        def __call__(self, data: Any, *, encoding: Encoding) -> RObject: # noqa: ANN401
-            """Convert Python object to R object."""
+        def convert_to_r_object(self, data: Any) -> RObject:  # noqa: ANN401
+            """Convert Python data to R object."""
 
+    ConstructorReturnValue = tuple[RObjectType, Any, dict[str, Any] | None]
+    ConstructorFunction1 = Callable[[Any], ConstructorReturnValue]
+    ConstructorFunction2 = Callable[[Any, Converter], ConstructorReturnValue]
```
Comment (vnmabus) [inline rdata/conversion/to_r.py]: For writing, as this is a new API, I think we can force the signature to always include the `Converter`. For reading we can't do that, as there are existing constructors without that parameter, and thus we would have to add the new signature and deprecate the old one.

File: rdata/conversion/to_r.py [hunk 21]
```diff
@@ -24,18 +27,31 @@
...
+
+        def convert_to_r_sym(self, name: str) -> RObject:
+            """Convert string to R symbol."""
 
-        def __call__(self, data: Any, *, encoding: Encoding) -> RObject: # noqa: ANN401
-            """Convert Python object to R object."""
+        def convert_to_r_object(self, data: Any) -> RObject:  # noqa: ANN401
+            """Convert Python data to R object."""
 
+    ConstructorReturnValue = tuple[RObjectType, Any, dict[str, Any] | None]
```
Comment (vnmabus) [inline rdata/conversion/to_r.py]: I am not sure if it wouldn't be better (less ambiguous and more flexible) to return the constructed R objects, even if that breaks symmetry with the reading constructors.

File: rdata/conversion/to_r.py (hunk 7 above)
Comment (vnmabus) [reply rdata/conversion/to_r.py:626]: I would add that as mandatory (at least for writing, as there is not existing code that requires backwards compatibility).

File: rdata/missing.py (hunk 9 above)
Comment (vnmabus) [reply rdata/missing.py]: I see.

File: rdata/missing.py (hunk 10 above)
Comment (vnmabus) [reply rdata/missing.py]: Understood.

**Review by vnmabus — CHANGES_REQUESTED** [review]

File: rdata/conversion/__init__.py (hunk 14 above)
Comment (trossi) [reply rdata/conversion/__init__.py]: Changed to `Utilities for converting between R and Python objects.`

File: rdata/conversion/to_r.py (hunk 20 above)
Comment (trossi) [reply rdata/conversion/to_r.py]: Done.

File: rdata/conversion/to_r.py (hunk 21 above)
Comment (trossi) [reply rdata/conversion/to_r.py]: I agree that would be better. Done.

File: rdata/_write.py (hunk 13 above)
Comment (trossi) [reply rdata/_write.py]: True, changed to `Dictionary mapping Python classes to functions converting them to R classes.` here and in other locations.

File: rdata/conversion/__init__.py (hunk 15 above)
Comment (trossi) [reply rdata/conversion/__init__.py]: I agree it is clearer to import this directly from `to_r`; removed from here. Regarding symmetric naming, do you mean we should rename it to `DEFAULT_CLASS_MAP` (same as in reading, user imports the correct name from `to_r` / `to_python`) or more explicit (but longer) like `DEFAULT_PYTHON_TO_R_CLASS_MAP`?

File: rdata/conversion/__init__.py (hunk 16 above)
Comment (trossi) [reply rdata/conversion/__init__.py]: Removed from here, but kept it in `to_r`as it's used only there.

File: rdata/conversion/to_r.py (hunk 18 above)
Comment (trossi) [reply rdata/conversion/to_r.py]: Removed these asserts that are already in type hints.

File: rdata/conversion/to_r.py (hunk 19 above)
Comment (trossi) [reply rdata/conversion/to_r.py:128]: This corresponds to default row names (indices 1,2,3,...) that are present in 'test_dataframe.rda' for example. I added a comment `# Construct default row names stored as [R_INT_NA, -len]`

File: rdata/conversion/to_r.py (hunk 4 above)
Comment (trossi) [reply rdata/conversion/to_r.py:231]: Sounds good. Maybe an issue could be opened about it for a reminder? (Relates to the comment about R_FLOAT_NA, np.nan, and other NaNs too.)

File: rdata/conversion/_conversion.py (hunk 17 above)
Comment (trossi) [reply rdata/conversion/_conversion.py:406]: Here is a demo for changing how R's NA vs NaN values are interpreted on pandas side.

This is for a file created `Rscript -e 'df = data.frame(a=c(1., NA, NaN)); saveRDS(df, file="na_nan.rds")'`.

```python
import numpy as np
import pandas as pd
import rdata
from rdata.missing import is_na

df = rdata.read_rds('na_nan.rds')
print('Dataframe df:')
print(df)
print()

print('df.isna():')
print(df.isna())
print()

print('np.isnan(df):')
print(np.isnan(df))
print()

print("Update NA mask of column")
a = df["a"].to_numpy()
df["a"] = pd.arrays.FloatingArray(a, is_na(a))
print()

print('New dataframe df:')
print(df)
print()

print('df.isna():')
print(df.isna())
print()

print('np.isnan(df):')
print(np.isnan(df))
print()
```

Output:
```
Dataframe df:
     a
1  1.0
2  NaN
3  NaN

df.isna():
       a
1  False
2   True
3   True

np.isnan(df):
       a
1  False
2   True
3   True

Update NA mask of column

New dataframe df:
      a
1   1.0
2  <NA>
3   NaN

df.isna():
       a
1  False
2   True
3  False

np.isnan(df):
       a
1  False
2   <NA>
3   True
```

It's a good question though what should be the default behavior in this regard. The current parsing behavior is the same as before this PR.

File: rdata/conversion/_conversion.py (hunk 2 above)
Comment (trossi) [reply rdata/conversion/_conversion.py:451]: The tests and mypy pass, so I kept this change.

File: rdata/conversion/_conversion.py (hunk 1 above)
Comment (trossi) [reply rdata/conversion/_conversion.py]: Now altrep is created only for suitable pd.RangeIndex objects. Users could avoid altreps here altogether by overriding the default constructor with

```python
def rangeindex_constructor(data: pd.RangeIndex, converter: Converter) -> RObject:
    return build_r_object(RObjectType.INT, value=np.array(data))
```

File: rdata/conversion/to_r.py (hunk 7 above)
Comment (trossi) [reply rdata/conversion/to_r.py:626]: Done.

**Review by trossi — COMMENTED** [review]
@vnmabus Sorry for the long delay with this PR - my past months were pretty busy. I have pushed fixes based on your suggestions.

File: rdata/conversion/__init__.py (hunk 15 above)
Comment (vnmabus) [reply rdata/conversion/__init__.py]: I think the shortest one is better. They are unlikely to be imported at the same time, and even that case is solvable with the `ìmport ... as` pattern.

Comment (trossi) [reply rdata/conversion/__init__.py]: I agree. Renamed `DEFAULT_CONSTRUCTOR_DICT` to `DEFAULT_CLASS_MAP`.

**Review by vnmabus — APPROVED** [review]
Thank you so much! I think this is ready. If there is something we missed, we can just change it later, I think.

**Comment by trossi** [comment]
@vnmabus Thank you for the thorough review once again!

It would be useful to have a pypi release with the writing functionality. Do you think the code would be ready for that? The writing functionality would need documentation in rdata.readthedocs.io (I can make a PR for that), but would something else be needed too?

**Comment by vnmabus** [comment]
> It would be useful to have a pypi release with the writing functionality. Do you think the code would be ready for that? The writing functionality would need documentation in rdata.readthedocs.io (I can make a PR for that), but would something else be needed too?

I would say that the only thing missing is the documentation.
